### PR TITLE
Add CRD Validation Rules to Beta blog post

### PR DESCRIPTION
**Pull request by jpbetz**
https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#beta

Blog post will explain the core capabilities of the feature, touch on some of the properties of the CEL expression language that was introduced by this feawture, explain what has changed in beta, and also describe new ways that CEL may be used in Kubernetes in the future.

### CODE REVIEW

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:3]:
```suggestion
title: "Kubernetes 1.25: CustomResourceDefinition Validation Rules Graduate to Beta"
```

**Comment by sftim** [comment]
/retitle [WIP] Add CRD Validation Rules to Beta blog post

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (rata) [reply content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:3]: s/Graduate/Graduates?

**Comment by jpbetz** [comment]
I've committed a 1st draft.  Review feedback much appreciated!

cc @cici37 @DangerOnTheRanger @alexzielenski @benluddy

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:6]:
```suggestion
```
We're not mirroring this article from elsewhere.

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:5]:
```suggestion
slug: crd-validation-rules-beta
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:11]: Ideally, link to the section about validation rules in the CRD concept docs (if there isn't a section yet, please mention the PR that'll add it so that a reviewer can cross-check).

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:45]:
```suggestion
Validation rules are expressive and flexible. See the [Validation Rules documentation](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules) to learn more about what validation rules are capable of.
```



Ideally, link to the section about validation rules in the CRD concept docs (if there isn't a section yet, please mention the PR that'll add it so that a reviewer can cross-check).

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:65]:
```suggestion
## Getting started with validation rules

### Writing validation rules in OpenAPIv3 schemas

```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:66]:
```suggestion
You can define validation rules for any level of a CRD's OpenAPIv3 schema. Validation rules are automatically scoped to their location in the schema where they are declared.
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:68]:
```suggestion
Good practices for CRD validation rules:
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:76]:
```suggestion
Examples of good practice:
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:85]:
```suggestion
## CRD transition rules
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:87]:
```suggestion
[Transition Rules](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#transition-rules) make it possible to compare the new state against the old state of a resource in validation rules. You use transition rules to make sure that the cluster's API server does not accept invalid state transitions. A transition rule is a validation rule that references 'oldSelf'. The API server only evaluates transition rules when both an old value and new value exist.
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:93]:
```suggestion
| `self == oldSelf` | For a required field, make that field immutable once it is set. For an optional field, only allow transitioning from unset to set, or from set to unset. |
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:94]:
```suggestion
| (on parent of field) `has(self.field) == has(oldSelf.field)`<br>on field:  `self == oldSelf` | Make a field immutable: validate that a field, even if optional, never changes after the resource is created (for a required field, the previous rule is simpler). |
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:95]:
```suggestion
| `self.all(x, x in oldSelf)` | Only allow adding items to a field that represents a set (prevent removals). |
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:106]: Should these links actually be to v0.25.0?

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:118]:
```suggestion
## Resource use and limits
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:120]:
```suggestion
To prevent CEL evaluation from consuming excessive compute resources, validation rules impose some limits. These limits are based on CEL _cost units_, a platform and machine independent measure of execution cost. As a result, the limits are the same regardless of where they are enforced.
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:122]:
```suggestion
### Estimated cost limit
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:124]:
```suggestion
CEL is, by design, non-Turing-complete in such a way that the halting problem isn’t a concern. CEL takes advantage of this design choice to include an "estimated cost" subsystem that can statically compute the worst case run time cost of any CEL expression. Validation rules are [integrated with the estimated cost system](o/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#resource-use-by-validation-functions) and disallow CEL expressions from being included in CRDs if they have a sufficiently poor (high) estimated cost. The estimated cost limit is set quite high and typically requires an O(n^2) or worse operation, across something of unbounded size, to be exceeded. Fortunately the fix is usually quite simple: because the cost system is aware of size limits declared in the CRD's schema, CRD authors can add size limits to the CRD's schema  (`maxItems` for arrays, `maxProperties` for maps, `maxLength` for strings) to reduce the estimated cost.
```

_aside_
The docs for the cost estimation system really belong within either https://kubernetes.io/docs/reference/ or https://kubernetes.io/docs/concepts/
Task pages are really for explaining what the reader must do to achieve an outcome, and therefore aren't the right place for longer explanations.

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:126]:
```suggestion
Good practice:
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:130]:
```suggestion
### Runtime cost limits for CRD validation rules
```

This helps with SEO by avoiding sending readers here when they want to limit the financial cost of running Kubernetes.

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:138]:
```suggestion
## Future work
```

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (sftim) [inline content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:140]:
```suggestion
We look forward to working with the community on the adoption of CRD Validation Rules, and hope to see this feature promoted to general availability in an upcoming Kubernetes release!
```

(graduation before 1.26? Unlikely :wink:)

**Review by sftim — COMMENTED** [review]
Thanks for the PR.

Here's some feedback; some of it is mechanical and there's a repeated aside where I call out the expectation of concept docs for CRD validation rules.

Also
/hold

Publication date is not agreed / assigned.

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (jpbetz) [reply content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:140]: I certainly won't be pushing for GA in 1.26.  Time to let the feature settle in for a bit.  Phrasing is nice here, thanks!

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (jpbetz) [reply content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:11]: Good catch. We don't have the section yet. I'll get that sorted out and fix this link.

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (jpbetz) [reply content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:106]: No, looks like I can omit the version part entirely. I'll fix.

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (jpbetz) [reply content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:124]: This is a really good point.  I feel like we've got a more general problem with CRD documentation here. I'll track it and start working on reorganizing it to fit the concept/task grouping.

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (jpbetz) [reply content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:130]: Nice.

File: content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md
Comment (jpbetz) [reply content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md:10]: This might be too much to fix in time for this blog entry, but I've opened https://github.com/kubernetes/website/issues/35991 to address the larger issue here.

**Comment by jpbetz** [comment]
Thanks for the feedback @sftim!  I've tried to apply it all-- I've accepted all the suggestions. Really appreciate the detailed review on this one.  Let me know if there is anything else I should do in the near term. Looks like we're otherwise just waiting for a publication date.

**Comment by fsmunoz** [comment]
Hello, Comms shadow for 1.25 here. This blog has been  scheduled for **September 23rd**, thought it would be good to share here. I'm also available for any assistance in the review process for this PR.

**Review by sftim — DISMISSED** [review]
Please update the publication date to the 23rd of September.

Other than that: it'd be good to get tech signoff (can be from a co-author, or from anyone else in the API Machinery SIG).

**Comment by cici37** [comment]
The context lgtm. Thanks for the work!
/lgtm

**Comment by sftim** [comment]
/lgtm cancel
Publication date is not right (in the committed code). Once that's fixed, OK to add LGTM.

File: content/en/blog/_posts/2022-09-23-crd-validation-rules-graduate-to-beta.md
Comment (Sea-n) [inline content/en/blog/_posts/2022-09-23-crd-validation-rules-graduate-to-beta.md:10]: We use relative path in k/website.
```suggestion
In Kubernetes 1.25, [Validation rules for CustomResourceDefinitions](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules) (CRDs) have graduated to Beta!
```

File: content/en/blog/_posts/2022-09-23-crd-validation-rules-graduate-to-beta.md
Comment (Sea-n) [inline content/en/blog/_posts/2022-09-23-crd-validation-rules-graduate-to-beta.md:59]: Please split into lines for long one.

```suggestion
* Cluster operability improves because CRD validation no longer requires a
  remote call to a webhook endpoint, eliminating a potential point of failure
  in the request-serving-path of the Kubernetes API server. This allows
  clusters to retain high availability while scaling to larger amounts of
  installed CRD extensions, since expected control plane availability would
  otherwise decrease with each additional webhook installed.
```

File: content/en/blog/_posts/2022-09-23-crd-validation-rules-graduate-to-beta.md
Comment (Sea-n) [inline content/en/blog/_posts/2022-09-23-crd-validation-rules-graduate-to-beta.md:141]:
```suggestion
There is a growing community of Kubernetes contributors thinking about how to make it possible to write extensible admission controllers using CEL as a substitute for admission webhooks for policy enforcement use cases. Anyone interested should reach out to us on the usual [SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery) channels or via Slack at [#sig-api-machinery-cel-dev](https://kubernetes.slack.com/archives/C02TTBG6LF4).
```

**Comment by sftim** [comment]
Merging this is better for the project than not merging.

/lgtm
/approve

**Pull request merged**
